src/Compilers/Test/Core/Compilation/CompilationExtensions.cs

```diff
@@ -381,6 +383,41 @@ void checkControlFlowGraph(IOperation root, ISymbol associatedSymbol)
             }
         }
 
+        internal static void VerifyChildNodePositions(this SyntaxTree tree)
+        {
+            var nodes = tree.GetRoot().DescendantNodesAndSelf();
+            foreach (var node in nodes)
+            {
+                var childNodesAndTokens = node.ChildNodesAndTokens();
+                if (childNodesAndTokens.Node is { } container)
+                {
+                    for (int i = 0; i < childNodesAndTokens.Count; i++)
+                    {
+                        if (container.GetNodeSlot(i) is Microsoft.CodeAnalysis.Syntax.SyntaxList.SeparatedWithManyChildren separatedList)
+                        {
+                            verifyPositions(separatedList);
+                        }
+                    }
+                }
+            }
+
+            static void verifyPositions(Microsoft.CodeAnalysis.Syntax.SyntaxList.SeparatedWithManyChildren separatedList)
+            {
+                int n = separatedList.SlotCount;
+                var positions1 = new int[n];
+                for (int i = 0; i < n; i++)
+                {
+                    positions1[i] = separatedList.GetChildPosition(i);
+                }
+                var positions2 = new int[n];
+                for (int i = n - 1; i >= 0; i--)
+                {
+                    positions2[i] = separatedList.GetChildPositionFromEnd(i);
+                }
```

Comment:
and just to check.  this actually computes, and is unaffected by anything potentially cached with the first for-loop?

Reply:
Good catch, thanks.

Actually, neither of the two loops cache any nodes, but both loops get the positions of the nodes using the cache is there at the time the verify method is called.

Add comments and also added verification without empty caches for both directions.